### DISCUSSION TITLE
MPD database loads but is inaccessible

### DISCUSSION THREAD

**Discussion by williemctell** [rern/rAudio#1394] [Bug Reports]
I've been having difficulty with a clean install of v 20230620 on a Raspberry Pi 3A+ using a known good micro SD card.  I got the database to load but can't access it from the web UI or more precisely, I can't see the files.   I get the "Database not yet available in this mode" error for files and the various tag views.  The database icon is flashing indicating an update in progress.  I am able to load and play music using mpc from the command line.  It shows up in the web UI now playing and the playlist/queue screen.  I have loaded a couple of directories using mpc ls piped to mpc add.  It's been about an hour since mpd.db had its last update.  My music files are on a NAS.  There are about 90K.  I've used Moode, Volumio, LMS, vanilla MPD, and Rune and never encountered this before.  I can also access the mpd database using MALP.  

Am I likely to be able to use the web UI to play files?

**Comment by klassikmann** [comment]
Perhaps your problem is related to #1377. If so, we have a regression.

> _Observing the update I see short calls to mpd, mpc and awk, then a cmd-list.sh using most of the time, and after that the search for .mpdignore files._

 I expect the problem will be solved with the next release.

**Reply by williemctell** [reply]
Just for the heck of it I tried installing Raudio on a Pi 4.  Same result.  journalctl for the end of the MPD job shows:

rAudio:~  #  journalctl _PID=1126 -f
Aug 29 17:12:56 rAudio mpd[1126]: decoder: Decoder plugin 'wildmidi' is unavailable: configuration file does not exist: /etc/timidity/timidity.cfg
Aug 29 17:12:58 rAudio mpd[1126]: exception: No such directory
Aug 29 17:12:58 rAudio mpd[1126]: exception: No such directory
Aug 29 17:12:58 rAudio mpd[1126]: client: error on client 11: Output buffer is full

MPD status shows:

rAudio:~  #  systemctl status mpd
* mpd.service - Music Player Daemon
     Loaded: loaded (/usr/lib/systemd/system/mpd.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/mpd.service.d
             `-00-arch.conf
             /etc/systemd/system/mpd.service.d
             `-override.conf
     Active: active (running) since Tue 2023-08-29 17:12:57 PDT; 20min ago
       Docs: man:mpd(1)
             man:mpd.conf(5)
   Main PID: 1126 (mpd)
      Tasks: 3 (limit: 4469)
        CPU: 2.033s
     CGroup: /system.slice/mpd.service
             `-1126 /usr/bin/mpd --systemd /srv/http/data/mpdconf/mpd.conf

Aug 29 17:12:55 rAudio systemd[1]: Starting Music Player Daemon...
Aug 29 17:12:56 rAudio mpd[1126]: decoder: Decoder plugin 'wildmidi' is unavailable: configuration file does>
Aug 29 17:12:57 rAudio systemd[1]: Started Music Player Daemon.
Aug 29 17:12:58 rAudio mpd[1126]: exception: No such directory
Aug 29 17:12:58 rAudio mpd[1126]: exception: No such directory
Aug 29 17:12:58 rAudio mpd[1126]: client: error on client 11: Output buffer is full

**Reply by williemctell** [reply]
Upgraded to 20230828 with the same results.  I can play audio using MPD directly and see the now playing file and the playlist in the rAudio web UI but the NAS contents remain unreachable.

**Comment by ceejayemm** [comment]
Hi.  From the ReadMe file for the site:

No albums found after update very large Library
Settings > MPD > Output Buffer - Increase by 8192 at a time
Update Library

Is your NAS by any chance a USB device attached to a FritzBox or similar router ?  If so there have been other instances of rAudio not liking connecting to such devices.  Check the forum for possible solutions - but if this is the case the best solution is to change to a 'proper' NAS that can provide NFS/CIFS shares 'properly'.

You don't say how you are mounting your NAS music share - NFS or CIFS (Samba) ?  Have you provided a Storage Name for the rAudio connection - see #1353 for discussion on this. rAudio REQUIRES that you provide a Storage Name for the music library share..

Chris

**Reply by williemctell** [reply]
My output buffer is set to 16K.  I'll try a bigger multiple.  

My NAS is a QNAP with SMB shares.  The router is a tplink AX3000.   The share is mounted as CIFS.  Yup, it has a human readable name in the rAudio config.  I like to name them after dead blues singers.  I set it up in the same way as I have in Moode, Volumio, LMS, and vanilla MPD.  My fstab entry is the standard one I use for MPD or another manual mount.  

The interesting thing to me is that I can play music from the mounted files using MPD either from the command line or from a phone app.  When I do that I can view what's playing in the rAudio browser UI.  I can see the file currently playing and the contents of the queue.  The problem is that rAudio reports that the database is incomplete so I can't search for files in the browser UI.  I don't normally use tags, but I did check the various tag views of the MPD database and got the same result as the file view.  The radio stations, stored in a different directory, are visible through the browser UI.

**Comment by ceejayemm** [comment]
Not sure what else to suggest. Are your music files encoded in any way in which case you might need to set Player / Options / FFmpeg to enabled ?

Chris

**Reply by williemctell** [reply]
No.  Most are vanilla FLAC with some mp3s.  The problem is that the NAS, even though the files are mounted and accessible to an MPD client, don't show up in the rAudio web UI.  I get a database not ready error.

**Reply by ceejayemm** [reply]
I think FFmpeg should be 'enabled' for FLAC files, shouldn't be needed for mp3 files but I believe it can be left on.

Chris

**Reply by williemctell** [reply]
Hmm...I'm able to play them without transcoding using an MPD client and see the now playing and queue in rAudio.  IIRC, MPD supports FLAC out of the box.  I'm assuming that like Moode, Volumio, et al rAudio uses MPD/MPC to play music.   

I'll try transcoding any way.  I was a programmer for 15 years and am used to trying things that I don't think will work.  I can't count the number of times that they solved the problem.

**Reply by klassikmann** [reply]
"mpd --version" shows the available decoders. Even if FFmpeg is not enabled in the rAudio player settings, ffmpeg will be decoded by the available flac decoder. Same argument applies to mp3 decoding.
I enabled FFmpeg to decode aac and ts in certain web radios.

**Comment by williemctell** [comment]
I'm not 100% sure but it appears that my output buffer size was the problem.  I went in 8K steps and eventually got the files to appear at 64K.  I don't know if this is an abnormally large buffer size.  I have about 90K files on the NAS.   The database load from scratch took about 5 hours which is comparable to other MPD front ends I've tried.

**Reply by williemctell** [reply]
It worked for about 10 minutes.  MPD stopped spontaneously and I couldn't restart it.  I'm going to try once more with the buffer set to 128 K.  I'm wondering what the upper limit on the buffer size is.

**Reply by williemctell** [reply]
Failed with larger buffer.  MPD timed out every time I tried to start it.  I'm wondering if I have too large a library although Moode, Volumio, LMS, and vanilla MPD have no problem loading it.

**Reply by klassikmann** [reply]
@williemctell 
If I interpret /srv/http/bash/cmd-list.sh (line 55 - 90) correctly, rAudio is automatically increasing max_output_buffer_size in 8192 steps.

**Reply by williemctell** [reply]
What I've seen is that below a hard coded 64 K the MPD database build completes with a buffer overflow error message.  I don't believe the buffer size increases automatically.  I should look at the script.  My current problem is the MPD load timeout.  I got the database build to complete with a 64K buffer but MPD crashed repeatedly over a period of 10 minutes or so.  For the first 5 or 6 times it reloaded briefly but after that, even after rebooting, it failed to start.

**Comment by williemctell** [comment]
Just for the lheck of it I'm trying to load my library in pieces.  My main share has a directory structure like:  /audio/genre/artist .  I loaded the second largest directory first and have what looks like success.  I'm currently adding the second largest directory.  If it works I'll have a bunch of mounts.

**Comment by williemctell** [comment]
I discovered that that if I only loaded one genre directory, my second largest, It worked fine.  Then I tried importing a my largest directory as a separate unit.  The first one was IP address/top level share/genre 1 and the second one was  IP address/top level share/genre 2.  It looks like rAudio can handle only one NAS share.  When the larger directory, genre 2, loaded the smaller one became invisible.  

I'm thinking that the size of the MPD database must be a limiting factor.  I have about 90K files.

**Comment by williemctell** [comment]
I reloaded my files Saturday night.  I set the output buffer size at 64 K as I had before.  Yesterday morning I tried it.  It worked. I listened for several hours and it continued to work. When I worked in IT we called things like this where we repeated something exactly the result of magic or sunspots.
